File: JavaScript/36-Valid Sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    let row = new Array(), col = new Array(), box = new Array();
    for (let i=0; i<9; i++){
        row.push(new Set());
        col.push(new Set());
        box.push(new Set());
    }
    
    // Don't fill the array as follows:
    // let row = new Array(9).fill(new Set());
    // let col = new Array(9).fill(new Set());
    // let box = new Array(9).fill(new Set());
    
    for (let r=0; r<9; r++){
        for (let c=0; c<9; c++){
            let curr = board[r][c];
            if (curr === ".") continue;
            let boxNum = Math.floor(r/3)*3 + Math.floor(c/3)
            if (row[r].has(curr) || col[c].has(curr) || box[boxNum].has(curr)){
                return false;
            }
            row[r].add(curr);
            col[c].add(curr);
            box[boxNum].add(curr);
        }
    }
    return true;
};

File: C++/36-Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<vector<bool>> row(9, vector<bool>(9, false));
        vector<vector<bool>> col(9, vector<bool>(9, false));
        vector<vector<bool>> block(9, vector<bool>(9, false));

        for (int r=0; r<9; r++) {
            for (int c=0; c<9; c++) {
                if (board[r][c] != '.') {
                    if (!row[r][board[r][c] - '0'-1] && !col[c][board[r][c] - '0'-1] && !block[(r/3)*3 + (c/3)][board[r][c] - '0'-1]) {
                        row[r][board[r][c] - '0'-1] = true;
                        col[c][board[r][c] - '0'-1] = true;
                        block[(r/3)*3 + (c/3)][board[r][c] - '0'-1] = true;
                    } else {
                        return false;
                    }
                }
            }
        }

        return true;
    }
};